### PR TITLE
 feat: KAN-319 Create GitHub JavaScript Actions for Slack notifications script

### PR DESCRIPTION
**Pull request by andriiholubenko**

### CODE REVIEW

File: .github/actions/slack-notify/src/index.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
+const { setOutput, getInput, setFailed } = require("@actions/core");
+const github = require("@actions/github");
+
+function date_diff_func(start, end) {
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:4]: in JS we should name functions using camel case – `dateDiffFunc`. The function name should describe what's happening here. `date_diff_func` doesn't describe that

File: .github/actions/slack-notify/src/index.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  var duration = end - start;
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:5]: please, change `var` to `const` or `let`.

File: .github/actions/slack-notify/action.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+inputs:
+  gh-token: # GitHub token
+    description: "GitHub token"
+    required: false
+    default: ${{ github.token }}
+outputs:
+  slack-message:
+    description: Slack notify message
+runs:
+  using: "node16"
```
Comment (vtereshyn) [inline .github/actions/slack-notify/action.yml:13]: why not 16+?

File: .github/actions/slack-notify/src/index.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      return v + text;
+    }
+  };
+  return (
+    format_func(days, "d", true) +
+    format_func(hours, "h", true) +
+    format_func(minutes, "m", true) +
+    format_func(seconds, "s", false)
+  );
+}
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:28]:
```suggestion
}

```

File: .github/actions/slack-notify/src/index.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+function status_icon_func(s) {
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:29]:
```suggestion
function getStatusIcon(s) {
```

File: .github/actions/slack-notify/src/index.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+      return ":no_entry:";
+    case "w_cancelled":
+      return ":warning:";
+    case "success":
+      return "\u2705";
+    case "failure":
+      return "\u274C";
+    default:
+      return "\u20e0";
+  }
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:43]: let's change this function and `status_text_func` to object mapping. Example:

```
const STATUS_TO_ICON = {
	failure: "\u20e0",
	success: "\u2705"
}
```

File: .github/actions/slack-notify/src/index.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    );
+
+    ////////////////////////////////////
+    // build slack notification message
+    ////////////////////////////////////
+
+    // build the message
+    var fields = [];
+    var is_wf_success = true;
+    var is_wf_failure = false;
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:86]: const or let

File: .github/actions/slack-notify/src/index.js [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+      });
+    }
+    var workflow_status = "w_cancelled";
+    if (is_wf_success) {
+      workflow_status = "w_success";
+    } else if (is_wf_failure) {
+      workflow_status = "w_failure";
+    }
+
+    var slack_msg = {
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:117]: please, do not use `var`

**Review by vtereshyn — CHANGES_REQUESTED** [review]

File: .github/actions/slack-notify/action.yml (hunk 3 above)
Comment (andriiholubenko) [reply .github/actions/slack-notify/action.yml:13]: image

Currently only Node12 and Node16 are available

File: .github/actions/slack-notify/package-lock.json [hunk 9]
```diff
@@ -0,0 +1,201 @@
+{
```
Comment (vtereshyn) [inline .github/actions/slack-notify/package-lock.json:1]: why are you using `npm` here?

File: .github/actions/slack-notify/package.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  "main": "src/index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "AstrumU",
+  "license": "ISC",
+  "dependencies": {
+    "@actions/core": "^1.10.0",
+    "@actions/github": "^5.1.1"
```
Comment (vtereshyn) [inline .github/actions/slack-notify/package.json:14]:
```suggestion
    "@actions/core": "1.10.0",
    "@actions/github": "5.1.1"
```

File: .github/actions/slack-notify/src/index.js [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  delta -= minutes * 60;
+  const seconds = Math.floor(delta % 60);
+  const formatFunc = function (v, text, check) {
+    if (v <= 0 && check) {
+      return "";
+    } else {
+      return v + text;
+    }
+  };
+  return (
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:22]: please, add spaces

File: .github/actions/slack-notify/src/index.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    w_cancelled: ":warning:",
+    success: "\u2705",
+    failure: "\u274C",
+  };
+
+  if (s in statusIcons) {
+    return statusIcons[s];
+  }
+
+  return "\u20e0";
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:43]:
```suggestion
  return statusIcons[s] || "\u20e0";
```

File: .github/actions/slack-notify/src/index.js [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    w_cancelled: "Warning",
+    success: "Success",
+    failure: "Failure",
+  };
+
+  if (s in statusText) {
+    return statusText[s];
+  }
+
+  return "\u20e0";
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:59]: same as above

**Review by vtereshyn — CHANGES_REQUESTED** [review]

File: .github/actions/slack-notify/src/index.js [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+  // format the duration
+  let delta = duration / 1000;
+  const days = Math.floor(delta / 86400);
+  delta -= days * 86400;
+  const hours = Math.floor(delta / 3600) % 24;
+  delta -= hours * 3600;
+  const minutes = Math.floor(delta / 60) % 60;
+  delta -= minutes * 60;
+  const seconds = Math.floor(delta % 60);
+  const formatFunc = function (v, text, check) {
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:15]: please, define this function outside. also there is no need to assign function to variable:

```
function formatFunc (v, text, check) {
....
....
```

File: .github/actions/slack-notify/src/index.js [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    const octokit = github.getOctokit(ghToken);
+
+    const wfRun = await octokit.rest.actions.getWorkflowRun({
+      owner: github.context.repo.owner,
+      repo: github.context.repo.repo,
+      run_id: github.context.runId,
+    });
+
+    const jobsResponse = await octokit.request(
+      wfRun.data.jobs_url + "?per_page=100"
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:68]:
```suggestion
      `${wfRun.data.jobs_url}?per_page=100`
```

File: .github/actions/slack-notify/src/index.js [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+          },
+        },
+      ],
+    };
+
+    const maxFields = 10;
+
+    const fieldsBlock = {
+      type: "section",
+      fields,
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:139]:
```suggestion
      fields
```

File: .github/actions/slack-notify/src/index.js [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+const github = require("@actions/github");
+
+const MAX_FIELDS = 10;
+
+function formatTime(time, timeType, check = true) {
+  if (time <= 0 && check) {
+    return "";
+  } else {
+    return time + timeType;
+  }
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:11]:
```suggestion
  if (time <= 0 && check) {
    return "";
  }

    return time + timeType;

```

File: .github/actions/slack-notify/src/index.js [hunk 18]
```diff
@@ -0,0 +1,150 @@
+const { setOutput, getInput, setFailed } = require("@actions/core");
+const github = require("@actions/github");
+
+const MAX_FIELDS = 10;
+
+function formatTime(time, timeType, check = true) {
+  if (time <= 0 && check) {
+    return "";
```
Comment (vtereshyn) [inline .github/actions/slack-notify/src/index.js:8]: what is this format about? it returns an empty string. do you need this?

Comment (andriiholubenko) [reply .github/actions/slack-notify/src/index.js:8]: Returning an empty value in the "formatTime" function is necessary in order not to show zero values for  - days, minutes, hours in "dateDifference" function. At the same time we want to show zero for seconds if needed.

**Review by vtereshyn — APPROVED** [review]

**Pull request merged**